File: apps/mobile/lib/models/narrative_sleeve.dart
// Phase 96 D-14 — Dart mirror of backend NarrativeSleeve Pydantic schema
// (services/backend/app/schemas/narrative_sleeve.py).
//
// Optional response envelope on CoachChatResponse.narrative_sleeve. The
// hook digit-free linter (Plan 96-03 T1, services/backend/app/services/
// coach/narrative_sleeve_lint.py) has already swapped any digit-bearing
// hook to the FR fallback before this struct is serialised — the Dart
// renderer treats the hook field as ground truth.
class NarrativeSleeve {
  /// 1-line attention-grabbing opener, digit-free at runtime per D-16.
  final String hook;

  /// Main body. Cited numbers have already been substituted via the
  /// Phase 94 citation gate + Phase 95 W2 pack double-lookup.
  final String caption;

  /// Verb-first call-to-action, ≤12 words at lint time. Rendered with
  /// a leading « › » glyph per UI-SPEC §NarrativeSleeve.
  final String nextStep;

  /// Static TOML lookup result by archetype × canton × life_event.
  /// Empty string when no entry matches — UI conditionally hides the
  /// metaphor block.
  final String metaphor;

  const NarrativeSleeve({
    required this.hook,
    required this.caption,
    required this.nextStep,
    this.metaphor = '',
  });

  /// Build from a JSON map. Backend emits camelCase keys via
  /// Pydantic v2 alias_generator=to_camel. Accepts both camelCase
  /// (`nextStep`) and snake_case (`next_step`) for robustness.
  factory NarrativeSleeve.fromJson(Map<String, dynamic> json) {
    final next = json['nextStep'] ?? json['next_step'] ?? '';
    return NarrativeSleeve(
      hook: (json['hook'] as String?) ?? '',
      caption: (json['caption'] as String?) ?? '',
      nextStep: next as String,
      metaphor: (json['metaphor'] as String?) ?? '',
    );
  }

  Map<String, dynamic> toJson() => <String, dynamic>{
        'hook': hook,
        'caption': caption,
        'nextStep': nextStep,
        'metaphor': metaphor,
      };
}

File: services/backend/app/schemas/narrative_sleeve.py
"""Phase 96 D-14 — NarrativeSleeve response envelope.

4-field contract :
- ``hook`` — 1-line attention-grabbing opener. MUST be digit-free at
  runtime ; the response middleware (lands in Plan 96-03 per D-16)
  swaps the hook to a generic fallback on `\\d` match. Pydantic does
  NOT enforce the digit-free invariant here because Pydantic raising
  would 500 the response — the middleware swap is the chosen safety
  surface.
- ``caption`` — main body ; the Phase 94 citation gate has already
  applied (numbers wrapped in ``{{cite:<key>}}`` then substituted via
  the Phase 95 W2 double-lookup at
  ``citation_parser._substitute_placeholders(*, pack=)``).
- ``next_step`` — verb-first call-to-action, ≤12 words at lint time
  (the word-count + verb-first linter lands in Plan 96-03 ; this
  schema enforces only the byte-length cap).
- ``metaphor`` — static TOML lookup result by
  ``archetype × canton × life_event`` from
  ``apps/mobile/assets/metaphors.toml`` (loaded at app boot in W1).
  Empty string when no entry matches.

Phase 96 W2 lands the contract (this file) + the additive optional
``CoachChatResponse.narrative_sleeve`` field. The response middleware
that enforces hook-digit-free + next_step word-count lands in
Plan 96-03 (W3) per D-16. Frozen + extra-forbid match the
``GroundingPackEntry`` precedent at
``services/backend/app/services/coach/grounding_pack.py:51``.
"""
from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field


class NarrativeSleeve(BaseModel):
    """4-field response envelope per CONTEXT D-14."""

    model_config = ConfigDict(frozen=True, extra="forbid")

    hook: str = Field(..., min_length=1, max_length=200)
    caption: str = Field(..., min_length=1, max_length=2000)
    next_step: str = Field(..., min_length=1, max_length=120)
    metaphor: str = Field(default="", max_length=200)
